doc/man/z80-asm.1
.TH Z80-ASM 1 "18 Apr 2018" "Z80-ASM Version 2.4" "Z80 Assembler"

.SH NAME
z80-asm \- command-line Z80 CPU assember for Unix and DO$

.SH SYNOPSIS
.B z80-asm [-wlhcf XX] inputfile[.asm] [startaddress[[:length] :outputfile] ]

.B z80-asm [-wlhcf XX] inputfile startaddress[:length]:outputfile ...

.SH DESCRIPTION
z80-asm is a two pass compiler of the Z80 CPU assembly language. It's first
argument is name of an input (source) file. If no sections follow no code is
generated.
In the case of exactly one section follows and the inputfile name ends with
.I .asm
the
.I outputfile
with its leading colon delimiter and the optional
.I length
may be omitted. Then the output is written to a file named
.I inputfilename.z80
Generally each section tells the compiler to save code to file 
.I outputfile
from address
.I start address
and with len
.I length.
Format of the file is described in
.B z80-file(5).

If an obligate
.I outputfile
is missing compilation is aborted after this section but error checking of the
code is done.
If
.I length
is ommited, compiler saves code from start address to furthest byte stored to
memory (inclusive). 
.PP
All addresses are given decimal in range from 0 to 65535 or hexadecimal in the range
from 0x0000 to 0xffff.
.PP
For assembler syntax see section
.B z80-asm(5).

.SH OPTIONS
Compiler called with no options or 
.I -h
option prints usage. Option
.I -w
turns on warning messages, when overwriting code in memory. Option
.I -l
writes a source listing in pass 2 to stdout. Each input line is prepended
by PC, generated machine code and line no of 5 digits. Option
.I -f
which must be followed by a 2 digit hexadecimal value XX sets the undefined
bytes in gaps created by DEFS and ORG statements to the given fill value XX.
Finally option
.I -c
prints a cross reference listing of the labels to stdout in case of no errors.

.SH BUGS
No bugs are known. :-) If you find some, please feel free to send me a bugreport.

.SH AUTHORS
Z80-ASM was written by Brainsoft <brain@artax.karlin.mff.cuni.cz> and Achim Flammenkamp <achim@math.uni-bielefeld.de>

.SH COPYING
Copying of this software must be under terms of the GNU General Public
License. There's absolutely no warranty on this software. For more details see
file
.B doc/COPY

.PP
(c) 1999-2004 Brainsoft

.SH SEE ALSO
z80-asm(5), z80-mon(1), z80-file(5)


doc/CHANGELOG
changes from version ?? to 1.1:
 * user interface add key  =  to change the registers
   A,B,C,D,E,H,L,IX,IY,A',B',C',D',E',H',L',I,R
 * display and update the cycle counter and the refresh register
 * BUG in disasm.c  f_ld  LD rr,(nn) [only 1 byte loaded]   fixed!
 * BUG in disasm.c  f_ex  EX (SP),IX break missing ==> runs into EX (SP),IY;   fixed!
 * EI replaced by IFF1 & IFF2 (correct LD A,R & LD A,I)
 * user interface  @(CPU reset), #(clear memory), *(clear all cpu register),
                   $(clear tick counter), &(SP init)
 * console.c:  BUG in c_init()  usuage of undefined variable t   fixed!

changes from version 1.1 to 1.9:
 * IN, OUT, INI, OUTD, ...  all implemented
 * console.c:  BUG(?) in c_init()  if ICANON cleared  MIN should bet set to 1
 * BUG in disasm.c  f_add  all 16-bit addition with carry bit :(   fixed!
 * declaration of an assembler label may be appended by a colon 
 * if PC, SP, and MEMP input is requested current values are displayed in hexa
 * HALT really halts the CPU now if interrupt is disabled else set runmode false
 * user interface  !(NMI) implemented
 * user interface  .  puts compiled instructions into memory and increases addr
 * user interface  = =  to change 16-bit registers BC,DE,HL,BC',DE',HL',IX,IY
 * BUG in disasm.c  f_adc  wrong carry/h/v flag calculation    fixed!
 * BUG in disasm.c  f_sbc  wrong carry/h/v flag calculation    fixed!
 * interrupt behaviour simulated, except IM=0 (probably tick count wrong)
 * user interface  D(dump/disassemble into file)  toggle DI/EI to key ^ changed
 * PC and SP defaults into hexa,  CALL, JP and RST in disassembling format, too
 * user interface  " (protocol instructions executed)
 * bank switching logic implemented (still experimental)
 * BUG in disasm.c  f_ld  if 16 bit load from 0xffff adress OVERFLOW   fixed!
 * disasm.c:  f_sll  opcode CB30-CB37  implemented
 * warnings and error_msg now  occupy only the very last line (row 23)
 * user interface  %  sets a breakpoint with a run counter at current PC
 * breakpoints (address & counter) are displayed in row 22
 * pseudo instruction EQU in assembler (asm.c) implemented 
 * CPU Halt-pin/status is displayed in upper right corner if active
 * turbo mode improved a lot: 4 levels now (16 min, 380 sec, 38 sec, 2.5 sec)
 * disasm.c:  f_nop2  all undocumented ED-opcodes implemented
 * all the man pages updated
 * turbo mode replaced by controling the clock speed/frequency
 * the ports are implemented as a map onto one binary file named .Z80_ports
 * hardware_clock.c  and  lcd_display.c  provided as examples for port usage

changes from version 1.9 to 2.0beta:
 * BUG in inst-dist.c  ED63  ld hl,(nn)  A_NUM -> A_PODDLE_NUM   fixed!
 * z80-asm permits in the case of exactly 1 section and inputfilename ends with
   .asm the outputfile to be optional (.asm is exchanged to .z80 for outputfile)
 * BUG in z80-asm: if start > highest byte in memory undefined outputfile length
 * HALT halts the CPU now if interrupt is enabled else set runmode false
   necessary for emulator mode ==>  user interface  &  sets a DI on stack
 * user interface  !  toggle keyboard read disabled/enabled in run mode
   user interface ESC now simulates a NMI
 * z80-mon has option -E for emulator mode (immediately CPU start)
 * SIGTERM, SIGHUP terminates z80-mon, SIGUSR1 causes reset, SIGUSR2 causes NMI,
   SIGABORT causes cpu_wait=1 and SIGTRAP makes a CPU dump
 * keyboard mapping implemented (used for IN from port stdin)
 * simul.c:  decode() now handles stray FD and DD prefixes
 * all FD and DD are simulated by decode() ==>  IXh,IXl,IYh,IYl introduced in
   regs.c and regs_token and set N_REGISTERS to 21 , global _ushort ddfdcb_reg
 * decoder and executer [decode() and  execute.c] totally checked. Now memory
   read access separated from memory write access. Usage of DATA pins!
 * in disasm.c: all 11 CB-executions and all 8 arith/logic-A-executions bundled
 * BUG in disasm.c: f_cp/f_cpd/f_cpi/f_cpdr/...  h-flag wrong calculated  fixed!
 * BUG in disasm.c: f_dec  p-flag wrong calculated  fixed!
 * the f_... functions in execute.c now return a pointer to char, doesn't print
 * total memory access encapsulated in memory.c / io-data access simulated
 * Makefile updated (*.h) goes now into hardware,  make install added
 * doc/man pages updated
 * in asm.c: write_to_memory(), set_start_address(), get_current_address()
   introduced for usage of compile() [now returns the number of bytes writen]

changes from version 2.0b to 2.0c:
 * BUG in asm.c: lexical_analysis: multiple label definition possible  fixed!
 * BUG in compile.c: c_add: ADD IY,IX valid(typo), ADD IY,IY invalid  fixed!
 * BUG in execute.c: f_alu: ADD,ADC,SBC use second operand (bad nemonic) fixed!
 * BUG in execute.c: f_neg: carry-flag always effected (carry=!!A)  fixed!
 * BUG in execute.c: f_bit: sign-flag effected only if bit 7 tested  fixed!
 * BUG in execute.c: f_daa: total nonsense! (>=4 independent errors)  fixed!
 * z80-asm.c Option -l added ==> in asm.c LISTING in pass 2 evaluated
 * in execute.c: in disassembling mode jr and djnz instruction offset with sign
 * user interface  t  enables base switch for disassembling constants
 * user interface  L  now can read true binary files into memory
 * user interface  a,b  changed to  u,v  and  e  deleted [reserved for register]
 * pseudo instructions  DEFM, ALIGN  in assembler (asm.c) implemented (sll,too)
 * Z80-Asm accepts integers to base 16 also if they have a h or H postfix
 * hash table structur/allocation optimized and option -c in z80-asm.c added

changes from version 2.0c to 2.0:
 * pseudo instruction  END  added and max label length to be 63 documented
 * in asm.h: better interface (highest_address,generated_bytes,set_compile_pass)
 * in asm.c: compile(void) now detects an EOF (take_line reads exactly 1 line)

changes from version 2.0 to 2.1
 * in asm.c: compile(char *txt) ==> take_line(..) not longer obligate!

z80-asm.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "hash.h"
#include "asm.h"
#include "file.h"

static FILE *input;

static unsigned char memory[MAX_MEM];
static unsigned short start,length;


unsigned char
write_to_memory(unsigned short index, unsigned char a)
{
 unsigned char previous=memory[index];
 memory[index]=a;
 return previous;
}


static int
take_line(char *line, int max_chars)
{
 if (!fgets(line,max_chars-1,input)){line[0]=0;return 1;}  /* end of file */
#ifdef UNIX
 if (strlen(line) && line[strlen(line)-1]=='\n')
#endif
 line[strlen(line)-1]=0;
#ifdef UNIX  /* reading DOS-files */
 if (strlen(line) && line[strlen(line)-1]=='\r')
    line[strlen(line)-1]=0;
#endif
#ifdef DOS
 if (!strlen(line))take_line(line,max_chars);
#endif
 return 0;
}


static void
usage(char *myname)
{
printf(
"Z80 assembler V 2.4.1 "
"(c)1999-2004 Brainsoft  (Copyleft) 1999-2018\n"
"Usage: %s [-w] [-h] [-l] [-f xx] [-c] <input file .asm> [<start address>[[:<length>]:<output file>]]\n"
"Usage: %s [-w] [-h] [-l] [-f xx] [-c] <input file> [<start address>[:<length>]:<output file>] ...\n"
,myname,myname);
}


void
error(int l,char *line,char *txt)
{
fprintf(stderr,"%s\nline %d: %s\n",line,l,txt);
}


/* return value: 0=OK, 1=syntax error */
/* parses command line options */

static int
parse_arg(char *arg,char **filename)
{
   unsigned long a;
   char *p, *q, *e;
 
   q=arg;
   if (!*q && !*filename)
   {  fprintf(stderr,"Error: Output filename not specified.\n");
      return 1;
   }
   for (p=arg ; (*p)&&(*p)!=':' ; p++);  /* find ':' or end of string */
   if (*p)
   {  *p=0;
      p++;
      a=strtoul(arg,&e,16);
      if (*e)
      {  fprintf(stderr,"Error: Starting address is not a number.\n");
         return 1;
      }
      if (a >= 1<<16)
      {  fprintf(stderr,"Error: Starting address out of range.\n");
         return 1;
      }
      start=(unsigned short)a;
      q=p;
   }
   else
      start=0;
   for (; (*p)&&(*p)!=':' ; p++);  /* find ':' or end of string */
   if (*p) /* start:length:filename */
   {
      *p=0;
      p++;
      a=strtoul(q,&e,10); 
      if (*e)
      {  fprintf(stderr,"Error: Length is not a number.\n");
         return 1;
      }
      if (a>=1<<16)
      {  fprintf(stderr,"Error: Length out of range.\n");
         return 1;
      }
      length=(unsigned short)a;
      if ((int)start+length >= 1<<16)
      {  fprintf(stderr,"Error: File is too long.\n");
         return 1;
      }
      q=p;
   }
   else
      length=0;  /* start:filename */
   if (*q)
      *filename=q;
   return 0;
}


struct info{ char *label; int value; unsigned lineno; };

static int compare(const struct info *left, const struct info *right)
{
   return  strcmp(left->label,right->label);
}   

int
main(int argc, char **argv)
{
FILE *output;
char *txt=0;
char line[512];
int s,a=0,b,cross=0;
/* a is used as memory init value */

for (b=s=1;s<argc&&*argv[s]=='-';b++)
{
   if (!*(argv[s]+b))
      b=0, s++;
   else if (*(argv[s]+b)=='w')
   {  printf("Warnings switched on.\n");WARNINGS=1;
   }
   else if (*(argv[s]+b)=='l') LISTING=1;
   else if (*(argv[s]+b)=='c') cross=1;
   else if (*(argv[s]+b)=='f')
   {  if (s+1>=argc || 1!=sscanf(argv[++s],"%2x",&a))
         fprintf(stderr,"Error: option -f needs a hexadecimal argument\n");
      else
         b=0, s++;
   }
   else if (*(argv[s]+b)=='h') {usage(argv[0]);}
   else if (*(argv[s]+b)=='?') {usage(argv[0]);}
   else if (*(argv[s]+b)=='-') {s++;break;} /* in case filename equals option */
   else fprintf(stderr,"Error: unknown option -%s\n",argv[s]+b);
}

if (s == argc) return 0;
input=fopen(argv[s],"r");
if (!input)
{fprintf(stderr,"Error: can't open input file \"%s\".\n",argv[s]);return 1;}

disable_pseudo=0;
memset(memory,a,1<<16);
asm_init((unsigned char)a);

a=0;
set_compile_pass(1);
set_start_address(0);
while (!a && !take_line(line,511))
  a= compile(line);
if (a==8) a=0;
if (!a)
  a=check_cond_nesting();
if (!a)
 {
 if (fseek(input,0,SEEK_SET))
 {asm_close();fprintf(stderr,"can't rewind input file \"%s\".\n",argv[s]);return 2;}
 set_compile_pass(2);
 set_start_address(0);
 while (!a && !take_line(line,511))
   a= compile(line);
 }
if (a==8) a=0;
fclose(input);
if (s+1 == argc)
{asm_close();fprintf(stderr,"No code generated\n");return 0;}
if (!a)
{
   for (b=1+s;b<argc;b++)
   {
      txt=0;
      if (argc==2+s && strlen(argv[s]) >= 5 &&
          !strcmp(argv[s]+strlen(argv[s])-4,".asm"))
      {  txt=argv[s];
         sprintf(txt+strlen(argv[s])-3,"z80");
      }
      if (parse_arg(argv[b],&txt))
      {  asm_close();  return 1;  }
      output=fopen(txt,"wb");
      if (!output)
      {  asm_close();
         fprintf(stderr,"Error: Can't open output file \"%s\".\n",txt);
         return 1;
      }
      write_header(output,start);
      if (length || generated_bytes() && highest_address() >= start)
         fwrite(memory+start,1,length?length:highest_address()+1-start,output);
      fclose(output);
   }
}
if (!a && (LISTING||cross))
   printf("%u bytes code generated and %u labels defined\n",
          generated_bytes(),table_entries());
if (!a && cross && (b=table_entries()))
{ struct info  *ele;
  ele= malloc(b*sizeof(struct info));
  for(a=0;next_table_entry(&ele[a].label,&ele[a].value,&ele[a].lineno);a++);
  qsort(ele,b,sizeof(struct info),compare);
  printf("       Cross reference:\n");
  printf("      symbol value hexa  line\n");
  for (a=0;a<b;a++)
   printf("%12.12s %5d %4hx %5u\n",ele[a].label,ele[a].value,
          (unsigned short)ele[a].value,ele[a].lineno);
  free(ele);
  a=0;
}
asm_close();
return a;
}
